Multiple Toolbox Shell + Tests

chynu opened the pull request:
- Created skeletons (function names and JSdoc) for functions to implement multiple toolbox support.
 - Added test functions to make sure multiple toolbox functionality works.

Note: most of the code in this PR will likely go into either `project.js` or `toolbox.js` during refactoring.

Things not yet outlined (via skeleton code) to do/test in future:
 - Displaying toolboxes in tree view.
 - Uploading toolbox to add to multiple toolboxes (will likely go directly into `WorkspaceFactoryController.importFile()`).

(`feature_addToolbox` is identical to `develop`.)

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/google/blockly-devtools/46)
<!-- Reviewable:end -->

chynu committed:
New branch for toolbox object

chynu committed:
New branch, created toolbox object

chynu committed:
New branch, created toolbox object

chynu committed:
working on toolbox class

chynu committed:
Adding multiple toolbox functionality

chynu committed:
Merge branch 'develop' into feat_addToolbox

chynu committed:
Adding multiple toolboxes

chynu committed:
Added category class

chynu committed:
saving and switching between toolboxes

chynu committed:
Merge branch 'feature_toolbox' of https://github.com/google/blockly-devtools into feature_toolbox

chynu committed:
Merge branch 'develop' into feat_addToolbox

chynu committed:
Multiple toolboxes working (create and save).

chynu committed:
Added functions to Toolbox class

chynu committed:
fixed renaming toolboxes

chynu committed:
Added JSDoc for multiple toolboxes, console.logs for debugging

chynu committed:
Debugging addToolbox JSUnit tests

chynu committed:
Bugfix for modal popup / scrollbar

chynu committed:
Debugging unit test functions

chynu committed:
Merge branch 'develop' into feature_toolbox

chynu committed:
Add toolbox class skeleton

chynu committed:
Refactoring for test-only PR

chynu committed:
Rearranging code for PR

chynu committed:
Reorganizing code to fit MVC

chynu committed:
Added JSDoc to test functions.

AnmAtAnm committed:
Merge pull request #47 from google/bugfix_zindex

Bugfix for modal popup / scrollbar

AnmAtAnm commented on src/workspacefactory/wfactory_controller.js:
```diff
@@ -102,6 +107,65 @@ WorkspaceFactoryController.MODE_JS = 'javascript';
+ *
+ * @param {string} originalName Original name of toolbox.
+ * @param {string} newName New name of toolbox.
+ * @returns {boolean} True if successfully renamed.
```
Because this may prompt for user input, this shouldn't return a success value immediately.  Consider a callback or Promise.

AnmAtAnm commented on src/workspacefactory/wfactory_controller.js:
```diff
@@ -45,9 +46,28 @@
+  // Currently displayed toolbox. Upon init, the current toolbox name is
+  // the default toolbox name, which is the empty string. Users are forced
+  // to rename this once they create multiple toolboxes.
+  this.currentToolbox = '';
```
Add type annotation to this new member variable.

(And if you don't mind, the other variables below.  But not a requirement.)

AnmAtAnm commented on src/workspacefactory/wfactory_controller.js:
```diff
@@ -102,6 +107,65 @@ WorkspaceFactoryController.MODE_JS = 'javascript';
+/**
+ * Saves XML currently in workspace into currently active toolbox under this.toolboxList.
+ *
+ * @returns {boolean} If saved successfully.
```
Because this may prompt for user feedback, the success may not be know immediately.  Consider using a callback or a Promise to return the success status.

AnmAtAnm commented on src/workspacefactory/wfactory_controller.js:
```diff
@@ -102,6 +107,65 @@ WorkspaceFactoryController.MODE_JS = 'javascript';
+ *
+ * @param {string} name Name of toolbox to display.
+ * @returns {boolean} If toolbox name was valid and displayed succesfully onto
+ *     workspace.
```
Because this may prompt for user feedback, the success may not be know immediately. Consider using a callback or a Promise to return the success status.

AnmAtAnm commented on src/workspacefactory/wfactory_controller.js:
```diff
@@ -402,6 +466,34 @@ WorkspaceFactoryController.prototype.exportJsFile = function(exportMode) {
+ *     toolbox (WorkspaceFactoryController.MODE_TOOLBOX) or preloaded workspace
+ *     (WorkspaceFactoryController.MODE_PRELOAD).
+ */
+WorkspaceFactoryController.prototype.exportAllXml = function(exportMode) {
```
Use a callback or Promise to signal completion and errors.

AnmAtAnm commented on src/workspacefactory/wfactory_controller.js:
```diff
@@ -102,6 +107,65 @@ WorkspaceFactoryController.MODE_JS = 'javascript';
+ * toolbox that the user will eventually rename if they add another toolbox.
+ */
+WorkspaceFactoryController.prototype.newToolbox = function() {
+  // TODO: implement
```
Maybe consider a console.log(..) call to say the unimplemented method has been called.  Good for testing the UI is doing what you expect.

Here and below.

AnmAtAnm commented on src/workspacefactory/wfactory_controller.js:
```diff
@@ -964,9 +1056,12 @@ WorkspaceFactoryController.prototype.importPreloadFromTree_ = function(tree) {
-WorkspaceFactoryController.prototype.clearAll = function() {
-  if (!confirm('Are you sure you want to clear all of your work in Workspace' +
+WorkspaceFactoryController.prototype.clearAll = function(conf) {
+  if (conf && !confirm('Are you sure you want to clear all of your work in Workspace' +
```
Adding the user confirmation means we should signal completion to the calling function, via callback or Promise.

AnmAtAnm reviewed (commented):
Partial review

rachel-fenichel commented:
Reviewed 7 of 8 files at r2.
Review status: all files reviewed at latest revision, 13 unresolved discussions.

---

*[index.html, line 174 at r2](https://reviewable.io:443/reviews/google/blockly-devtools/46#-KnGuBiPKZrUkK0GewwC:-KnGuBiPKZrUkK0GewwD:b-cj30yw) ([raw file](https://github.com/google/blockly-devtools/blob/5c284bc5d95c02d1344826ed42a9daf59740fddb/index.html#L174)):*
> ```HTML
>             <a id="dropdown_exportToolboxXML">Toolbox as XML</a>
>             <a id="dropdown_exportToolboxJS">Toolbox as JS</a>
>             <a id="dropdown_exportAllToolboxXML">All Toolobxes as XML</a>
> ```

speeeeling

---

*[src/workspacefactory/wfactory_controller.js, line 116 at r2](https://reviewable.io:443/reviews/google/blockly-devtools/46#-KnGZ8iCTz--UUltKJlO-r2-116:-KnGudfSZR__tu43b2Ra:b32v57) ([raw file](https://github.com/google/blockly-devtools/blob/5c284bc5d95c02d1344826ed42a9daf59740fddb/src/workspacefactory/wfactory_controller.js#L116)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

Maybe consider a console.log(..) call to say the unimplemented method has been called.  Good for testing the UI is doing what you expect.

Here and below.
</blockquote></details>

+1

---

*[src/workspacefactory/wfactory_controller.js, line 127 at r2](https://reviewable.io:443/reviews/google/blockly-devtools/46#-KnGZ8iCTz--UUltKJlO-r2-127:-KnGuhZsS5n4cFNipGXs:b-qqni1y) ([raw file](https://github.com/google/blockly-devtools/blob/5c284bc5d95c02d1344826ed42a9daf59740fddb/src/workspacefactory/wfactory_controller.js#L127)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

Because this may prompt for user feedback, the success may not be know immediately.  Consider using a callback or a Promise to return the success status.
</blockquote></details>

Use "Whether" instead of "If".  Right now it reads as "returns if saved successfully, otherwise ???"

---

*[src/workspacefactory/wfactory_controller.js, line 1060 at r2](https://reviewable.io:443/reviews/google/blockly-devtools/46#-KnGv2Y7DRtR62pextbN:-KnGv2Y7DRtR62pextbO:b-8twui7) ([raw file](https://github.com/google/blockly-devtools/blob/5c284bc5d95c02d1344826ed42a9daf59740fddb/src/workspacefactory/wfactory_controller.js#L1060)):*
> ```JavaScript
>  * "Clear" button.
>  *
>  * @param {boolean} conf Whether user should first confirm before clearing
> ```

`shouldConfirm` is clearer than `conf`

---

*[src/workspacefactory/wfactory_model.js, line 96 at r2](https://reviewable.io:443/reviews/google/blockly-devtools/46#-KnGvGLdGmdKHKR4Dbv5:-KnGvGLeDelSTXq7nSxw:b-frhe0p) ([raw file](https://github.com/google/blockly-devtools/blob/5c284bc5d95c02d1344826ed42a9daf59740fddb/src/workspacefactory/wfactory_model.js#L96)):*
> ```JavaScript
>  * @returns {boolean} If updated successfully.
>  */
> WorkspaceFactoryModel.prototype.updateToolbox = function(name, xml) {
> ```

Will you ever have ways to update it other than from XML?  If so, name this `updateToolboxFromXml`.

---

*[src/workspacefactory/wfactory_model.js, line 103 at r2](https://reviewable.io:443/reviews/google/blockly-devtools/46#-KnGvVBOXB9pqP6uE3jc:-KnGvVBOXB9pqP6uE3jd:b-wainja) ([raw file](https://github.com/google/blockly-devtools/blob/5c284bc5d95c02d1344826ed42a9daf59740fddb/src/workspacefactory/wfactory_model.js#L103)):*
> ```JavaScript
>  * Renames toolbox from oldName to newName. Catches for duplicates and invalid
>  * names (empty strings, etc.).
>  *
> ```

You will need to add annotations for the arguments here and below when you implement these functions.

---

*[tests/wfactory_input_test.js, line 150 at r2](https://reviewable.io:443/reviews/google/blockly-devtools/46#-KnGvuLyQyF4lZIVaA4P:-KnGvuLyQyF4lZIVaA4Q:bz7nhz8) ([raw file](https://github.com/google/blockly-devtools/blob/5c284bc5d95c02d1344826ed42a9daf59740fddb/tests/wfactory_input_test.js#L150)):*
> ```JavaScript
>   model.renameToolbox('', 'hello');
> 
>   assertTrue('',
> ```

If you're not providing a comment just use the single-argument version instead of passing in an empty string.

---

*[tests/wfactory_input_test.js, line 186 at r2](https://reviewable.io:443/reviews/google/blockly-devtools/46#-KnGw7chlligeyLH_z-U:-KnGw7chlligeyLH_z-V:b364qbd) ([raw file](https://github.com/google/blockly-devtools/blob/5c284bc5d95c02d1344826ed42a9daf59740fddb/tests/wfactory_input_test.js#L186)):*
> ```JavaScript
>   numToolboxes += 1;
> 
>   assertFalse('FAILED: Toolbox names should not be able to be renamed to a ' +
> ```

Break this test into multiple pieces (yes, you will be duplicating some code).

With a test like this there are certain failures that you won't know about until you fix the failures that occur earlier in the text, regardless of whether those failures are related.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-devtools/46)*
<!-- Sent from Reviewable.io -->

chynu committed:
Merge branch 'develop' into test_addToolbox

chynu committed:
Added Promises to toolbox functions; added tests.

chynu commented:
Review status: 0 of 8 files reviewed at latest revision, 13 unresolved discussions.

---

*[index.html, line 174 at r2](https://reviewable.io:443/reviews/google/blockly-devtools/46#-KnGuBiPKZrUkK0GewwC:-KnKq1RArxXpt0vVVx3F:b-u3kzeh) ([raw file](https://github.com/google/blockly-devtools/blob/5c284bc5d95c02d1344826ed42a9daf59740fddb/index.html#L174)):*
<details><summary><i>Previously, rachel-fenichel (Rachel Fenichel) wrote…</i></summary><blockquote>

speeeeling
</blockquote></details>

oops-- done!

---

*[src/workspacefactory/wfactory_controller.js, line 55 at r2](https://reviewable.io:443/reviews/google/blockly-devtools/46#-KnGZ8iCTz--UUltKJlO-r2-55:-KnKr3mZylf52xZd-Uru:b-896fja) ([raw file](https://github.com/google/blockly-devtools/blob/5c284bc5d95c02d1344826ed42a9daf59740fddb/src/workspacefactory/wfactory_controller.js#L55)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

Add type annotation to this new member variable.

(And if you don't mind, the other variables below.  But not a requirement.)
</blockquote></details>

Done!

---

*[src/workspacefactory/wfactory_controller.js, line 116 at r2](https://reviewable.io:443/reviews/google/blockly-devtools/46#-KnGZ8iCTz--UUltKJlO-r2-116:-KnKrX8px_Br4aoqVmU5:b-f3htak) ([raw file](https://github.com/google/blockly-devtools/blob/5c284bc5d95c02d1344826ed42a9daf59740fddb/src/workspacefactory/wfactory_controller.js#L116)):*
<details><summary><i>Previously, rachel-fenichel (Rachel Fenichel) wrote…</i></summary><blockquote>

+1
</blockquote></details>

Okay. Done!

---

*[src/workspacefactory/wfactory_controller.js, line 127 at r2](https://reviewable.io:443/reviews/google/blockly-devtools/46#-KnGZ8iCTz--UUltKJlO-r2-127:-KnKsjAZcwCS6rdeV7ZI:bk3tv9p) ([raw file](https://github.com/google/blockly-devtools/blob/5c284bc5d95c02d1344826ed42a9daf59740fddb/src/workspacefactory/wfactory_controller.js#L127)):*
<details><summary><i>Previously, rachel-fenichel (Rachel Fenichel) wrote…</i></summary><blockquote>

Use "Whether" instead of "If".  Right now it reads as "returns if saved successfully, otherwise ???"
</blockquote></details>

Ooooh, cool! I changed the return type to a Promise that either resolves with a boolean (true) or rejects with the error message (to prompt the user again). Not sure if resolving/rejecting with different types is a standard JS move (haven't used Promises before), though.

---

*[src/workspacefactory/wfactory_controller.js, line 141 at r2](https://reviewable.io:443/reviews/google/blockly-devtools/46#-KnGZ8iCTz--UUltKJlO-r2-141:-KnKv9D86ueKsgxNIORC:bypep8k) ([raw file](https://github.com/google/blockly-devtools/blob/5c284bc5d95c02d1344826ed42a9daf59740fddb/src/workspacefactory/wfactory_controller.js#L141)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

Because this may prompt for user feedback, the success may not be know immediately. Consider using a callback or a Promise to return the success status.
</blockquote></details>

Changed return type to promise!

---

*[src/workspacefactory/wfactory_controller.js, line 158 at r2](https://reviewable.io:443/reviews/google/blockly-devtools/46#-KnGZ8iCTz--UUltKJlO-r2-158:-KnKvBXHN2gFmmjOFVPQ:b-z97szz) ([raw file](https://github.com/google/blockly-devtools/blob/5c284bc5d95c02d1344826ed42a9daf59740fddb/src/workspacefactory/wfactory_controller.js#L158)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

Because this may prompt for user input, this shouldn't return a success value immediately.  Consider a callback or Promise.
</blockquote></details>

Changed return type to Promise.

---

*[src/workspacefactory/wfactory_controller.js, line 478 at r2](https://reviewable.io:443/reviews/google/blockly-devtools/46#-KnGZ8iCTz--UUltKJlO-r2-478:-KnLC5vh_Q8XUQdoo7Vu:b-gwjt9m) ([raw file](https://github.com/google/blockly-devtools/blob/5c284bc5d95c02d1344826ed42a9daf59740fddb/src/workspacefactory/wfactory_controller.js#L478)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

Use a callback or Promise to signal completion and errors.
</blockquote></details>

Hm, I'm not sure I understand why Promises would be useful in this case. Is it because a user might cancel while exporting the XML files?

---

*[src/workspacefactory/wfactory_controller.js, line 1060 at r2](https://reviewable.io:443/reviews/google/blockly-devtools/46#-KnGv2Y7DRtR62pextbN:-KnKvG5NXWiZvsQnfwUh:b-896fja) ([raw file](https://github.com/google/blockly-devtools/blob/5c284bc5d95c02d1344826ed42a9daf59740fddb/src/workspacefactory/wfactory_controller.js#L1060)):*
<details><summary><i>Previously, rachel-fenichel (Rachel Fenichel) wrote…</i></summary><blockquote>

`shouldConfirm` is clearer than `conf`
</blockquote></details>

Done!

---

*[src/workspacefactory/wfactory_controller.js, line 1064 at r2](https://reviewable.io:443/reviews/google/blockly-devtools/46#-KnGZ8iCTz--UUltKJlO-r2-1064:-KnKwZtoCD1e9o3UBPsT:b7fk7m0) ([raw file](https://github.com/google/blockly-devtools/blob/5c284bc5d95c02d1344826ed42a9daf59740fddb/src/workspacefactory/wfactory_controller.js#L1064)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

Adding the user confirmation means we should signal completion to the calling function, via callback or Promise.
</blockquote></details>

Added a return type {Promise} that rejects if user does not confirm.

---

*[src/workspacefactory/wfactory_model.js, line 96 at r2](https://reviewable.io:443/reviews/google/blockly-devtools/46#-KnGvGLdGmdKHKR4Dbv5:-KnKyCv-kor2QroA88kG:b-yklmsa) ([raw file](https://github.com/google/blockly-devtools/blob/5c284bc5d95c02d1344826ed42a9daf59740fddb/src/workspacefactory/wfactory_model.js#L96)):*
<details><summary><i>Previously, rachel-fenichel (Rachel Fenichel) wrote…</i></summary><blockquote>

Will you ever have ways to update it other than from XML?  If so, name this `updateToolboxFromXml`.
</blockquote></details>

Renamed, but not sure how else we would retrieve the blocks currently on the workspace (?).

---

*[src/workspacefactory/wfactory_model.js, line 103 at r2](https://reviewable.io:443/reviews/google/blockly-devtools/46#-KnGvVBOXB9pqP6uE3jc:-KnKzJkzOdSL1MF8_cv4:bg49q8k) ([raw file](https://github.com/google/blockly-devtools/blob/5c284bc5d95c02d1344826ed42a9daf59740fddb/src/workspacefactory/wfactory_model.js#L103)):*
<details><summary><i>Previously, rachel-fenichel (Rachel Fenichel) wrote…</i></summary><blockquote>

You will need to add annotations for the arguments here and below when you implement these functions.
</blockquote></details>

Oops.. done.

---

*[tests/wfactory_input_test.js, line 150 at r2](https://reviewable.io:443/reviews/google/blockly-devtools/46#-KnGvuLyQyF4lZIVaA4P:-KnL-As8T7NFfHCdCP5q:b-896fja) ([raw file](https://github.com/google/blockly-devtools/blob/5c284bc5d95c02d1344826ed42a9daf59740fddb/tests/wfactory_input_test.js#L150)):*
<details><summary><i>Previously, rachel-fenichel (Rachel Fenichel) wrote…</i></summary><blockquote>

If you're not providing a comment just use the single-argument version instead of passing in an empty string.
</blockquote></details>

Done!

---

*[tests/wfactory_input_test.js, line 186 at r2](https://reviewable.io:443/reviews/google/blockly-devtools/46#-KnGw7chlligeyLH_z-U:-KnL9WXeaNHquCq9kbEa:b-yohfhc) ([raw file](https://github.com/google/blockly-devtools/blob/5c284bc5d95c02d1344826ed42a9daf59740fddb/tests/wfactory_input_test.js#L186)):*
<details><summary><i>Previously, rachel-fenichel (Rachel Fenichel) wrote…</i></summary><blockquote>

Break this test into multiple pieces (yes, you will be duplicating some code).

With a test like this there are certain failures that you won't know about until you fix the failures that occur earlier in the text, regardless of whether those failures are related.
</blockquote></details>

I see. I broke it into multiple test functions (and added a couple).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-devtools/46)*
<!-- Sent from Reviewable.io -->

chynu committed:
Modify jsdoc.

chynu commented:
Aha.... t'was an accident.

rachel-fenichel commented:
Reviewed 1 of 8 files at r3, 6 of 7 files at r4, 1 of 1 files at r5.
Review status: all files reviewed at latest revision, 17 unresolved discussions.

---

*[tests/wfactory_input_test.js, line 134 at r5](https://reviewable.io:443/reviews/google/blockly-devtools/46#-KnLWeVXz-ElQee-2NSb:-KnLWeVXz-ElQee-2NSc:b-cwzetn) ([raw file](https://github.com/google/blockly-devtools/blob/311c5540046eeae10dd72a0786576b87c6f5e087/tests/wfactory_input_test.js#L134)):*
> ```JavaScript
>       (errorMsg) => {
>         // Rejected. Fail test.
>         assertTrue('FAILED: ' + errorMsg, false);
> ```

Call `fail('comment')` instead of `assertTrue('comment', false)`

---

*[tests/wfactory_input_test.js, line 160 at r5](https://reviewable.io:443/reviews/google/blockly-devtools/46#-KnLWxBw7ImS-4e1MHey:-KnLWxBw7ImS-4e1MHez:b-4xfyb7) ([raw file](https://github.com/google/blockly-devtools/blob/311c5540046eeae10dd72a0786576b87c6f5e087/tests/wfactory_input_test.js#L160)):*
> ```JavaScript
>       (errorMsg) => {
>         // Rejected. Fail test.
>         assertTrue('FAILED: ' + errorMsg, false);
> ```

Use `fail` here as well, and anywhere below that you're using this pattern.

---

*[tests/wfactory_input_test.js, line 171 at r5](https://reviewable.io:443/reviews/google/blockly-devtools/46#-KnLXCzrtbdKy6W-vBMk:-KnLXCzrtbdKy6W-vBMl:bg9otny) ([raw file](https://github.com/google/blockly-devtools/blob/311c5540046eeae10dd72a0786576b87c6f5e087/tests/wfactory_input_test.js#L171)):*
> ```JavaScript
>           // Resolved. Check if it exists in toolboxList.
>           if (!model.toolboxNameIsTaken(specialChars[i])) {
>             assertTrue('FAILED: Adding specialChar ' + charLiteral[i] +
> ```

Instead of 
```
if (!condition) {
   assertTrue('message', false);
}
```

You can do
```
assertTrue('message', condition);
```

---

*[tests/wfactory_input_test.js, line 179 at r5](https://reviewable.io:443/reviews/google/blockly-devtools/46#-KnLXS8OXPIAbj2X7qbJ:-KnLXS8OXPIAbj2X7qbK:b-3ze1c9) ([raw file](https://github.com/google/blockly-devtools/blob/311c5540046eeae10dd72a0786576b87c6f5e087/tests/wfactory_input_test.js#L179)):*
> ```JavaScript
>           // Rejected. Fail test.
>           assertTrue(
>               'FAILED [iteration ' + i + ']: Failed in renaming to special character "' +
> ```

This comment mentions renaming, but you're just adding.  (Unless I'm missing something).

---

*[tests/wfactory_input_test.js, line 184 at r5](https://reviewable.io:443/reviews/google/blockly-devtools/46#-KnLZFdAdAkJQ_MSAuTS:-KnLZFdAdAkJQ_MSAuTT:b-nxc48h) ([raw file](https://github.com/google/blockly-devtools/blob/311c5540046eeae10dd72a0786576b87c6f5e087/tests/wfactory_input_test.js#L184)):*
> ```JavaScript
>         });
>     assertEquals('FAILED: Toolbox has ' + len(controller) + 'toolboxes saved. ' +
>           'Should have ' + (i + 1) + ' toolboxes.',
> ```

As discussed offline, assertEquals comments just need to provide context--the test framework will provide the "expected A but got B" string.

---

*[tests/wfactory_input_test.js, line 330 at r5](https://reviewable.io:443/reviews/google/blockly-devtools/46#-KnLZzo806kB3ErDwa8p:-KnLZzo806kB3ErDwa8q:bl7hfym) ([raw file](https://github.com/google/blockly-devtools/blob/311c5540046eeae10dd72a0786576b87c6f5e087/tests/wfactory_input_test.js#L330)):*
> ```JavaScript
>       1, len(controller));
> 
>   model.addToolbox('Test').then(
> ```

Since you're doing this at the start of many of these tests, make a helper function.  Pass in `model` and have it do all of this chunk, including the assert.  
The rest of your test won't run if the assert fails.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-devtools/46)*
<!-- Sent from Reviewable.io -->

svouse committed:
Refactor (#48)

Beginning of basic refactoring

chynu committed:
Merge branch 'develop' into test_addToolbox

chynu committed:
Modified asserts, added helper function (#46)

chynu commented:
Review status: 1 of 16 files reviewed at latest revision, 17 unresolved discussions.

---

*[tests/wfactory_input_test.js, line 134 at r5](https://reviewable.io:443/reviews/google/blockly-devtools/46#-KnLWeVXz-ElQee-2NSb:-KnLgQKvVxEy_0T17ybD:b-896fix) ([raw file](https://github.com/google/blockly-devtools/blob/311c5540046eeae10dd72a0786576b87c6f5e087/tests/wfactory_input_test.js#L134)):*
<details><summary><i>Previously, rachel-fenichel (Rachel Fenichel) wrote…</i></summary><blockquote>

Call `fail('comment')` instead of `assertTrue('comment', false)`
</blockquote></details>

Done.

---

*[tests/wfactory_input_test.js, line 160 at r5](https://reviewable.io:443/reviews/google/blockly-devtools/46#-KnLWxBw7ImS-4e1MHey:-KnLhhmZTfBjYrCeELR1:b-896fix) ([raw file](https://github.com/google/blockly-devtools/blob/311c5540046eeae10dd72a0786576b87c6f5e087/tests/wfactory_input_test.js#L160)):*
<details><summary><i>Previously, rachel-fenichel (Rachel Fenichel) wrote…</i></summary><blockquote>

Use `fail` here as well, and anywhere below that you're using this pattern.
</blockquote></details>

Done.

---

*[tests/wfactory_input_test.js, line 171 at r5](https://reviewable.io:443/reviews/google/blockly-devtools/46#-KnLXCzrtbdKy6W-vBMk:-KnLi3ODRulRnB9AiJU6:b-896fix) ([raw file](https://github.com/google/blockly-devtools/blob/311c5540046eeae10dd72a0786576b87c6f5e087/tests/wfactory_input_test.js#L171)):*
<details><summary><i>Previously, rachel-fenichel (Rachel Fenichel) wrote…</i></summary><blockquote>

Instead of 
```
if (!condition) {
   assertTrue('message', false);
}
```

You can do
```
assertTrue('message', condition);
```
</blockquote></details>

Done.

---

*[tests/wfactory_input_test.js, line 179 at r5](https://reviewable.io:443/reviews/google/blockly-devtools/46#-KnLXS8OXPIAbj2X7qbJ:-KnLi8k5C7SxgvNh_DZI:b-okqc7d) ([raw file](https://github.com/google/blockly-devtools/blob/311c5540046eeae10dd72a0786576b87c6f5e087/tests/wfactory_input_test.js#L179)):*
<details><summary><i>Previously, rachel-fenichel (Rachel Fenichel) wrote…</i></summary><blockquote>

This comment mentions renaming, but you're just adding.  (Unless I'm missing something).
</blockquote></details>

Ah, downfall of copy pasta. Changed comment.

---

*[tests/wfactory_input_test.js, line 184 at r5](https://reviewable.io:443/reviews/google/blockly-devtools/46#-KnLZFdAdAkJQ_MSAuTS:-KnLiVN4nPBG5Y0MnJOM:b-896fja) ([raw file](https://github.com/google/blockly-devtools/blob/311c5540046eeae10dd72a0786576b87c6f5e087/tests/wfactory_input_test.js#L184)):*
<details><summary><i>Previously, rachel-fenichel (Rachel Fenichel) wrote…</i></summary><blockquote>

As discussed offline, assertEquals comments just need to provide context--the test framework will provide the "expected A but got B" string.
</blockquote></details>

Done!

---

*[tests/wfactory_input_test.js, line 330 at r5](https://reviewable.io:443/reviews/google/blockly-devtools/46#-KnLZzo806kB3ErDwa8p:-KnLjdKCBYNYtTSgCV8s:b-896fix) ([raw file](https://github.com/google/blockly-devtools/blob/311c5540046eeae10dd72a0786576b87c6f5e087/tests/wfactory_input_test.js#L330)):*
<details><summary><i>Previously, rachel-fenichel (Rachel Fenichel) wrote…</i></summary><blockquote>

Since you're doing this at the start of many of these tests, make a helper function.  Pass in `model` and have it do all of this chunk, including the assert.  
The rest of your test won't run if the assert fails.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-devtools/46)*
<!-- Sent from Reviewable.io -->

rachel-fenichel commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 11 of 18 files at r6, 7 of 7 files at r7.
Review status: all files reviewed at latest revision, 8 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-devtools/46#-:-KnLnj8Eu6TtUPZmkIUp:bnfp4nl)*
<!-- Sent from Reviewable.io -->
